### ISSUE TITLE
AbstractRecord.from(String[], TableField[]) no longer works as expected

### ISSUE THREAD

**Issue by alf** [jOOQ/jOOQ#13040]
### Expected behavior

While testing 3.17.0-SNAPSHOT for #12608 I believe I've stumbled upon a regression in `AbstractRecord.from(String[], TableField[])` that seems to have been introduced while solving #12704.

I have code as follows:

```
var fields = key.getFieldsArray();
String[] values = unpack(key, s);
rec.from(values, fields);
```

In jOOQ 3.15.5 this works as I expect and the record is populated properly. In jOOQ 3.17.0-SNAPSHOT my unit tests fails. Digging deeper I gather the problem is caused by [the following](https://github.com/jOOQ/jOOQ/blob/82a565f5665bb96d74a58c60c81a723c3c960afb/jOOQ/src/main/java/org/jooq/impl/AbstractRecord.java#L932):

```
    @Override
    public final void from(Object source, Field<?>... f) {
        from0(source, fields.fields.indexesOf(f));
    }
```

It seems jOOQ now assumes that source has the same shape as the record we're populating. This is not correct in my case. This might of course be me using the API wrong, but I doubt this is the case as `fromArray(Object[], Field<?>... f)` delegates to the method I'm calling and thus does not fix the bug.

**Comment by lukaseder** [comment]
Thanks a lot for the heads up. I will investigate this soon.

**Comment by lukaseder** [comment]
I can confirm this issue with the following test:

```java
@Test
public void testFromArrayExplicitFields() {
    Record3<Integer, Integer, Integer> r = create.newRecord(FIELD_ID1, FIELD_ID2, FIELD_ID3);

    r.from(new Integer[] { 1 }, FIELD_ID2);
    assertEquals(asList(null, 1, null), r.intoList());

    r.from(new Integer[] { 1, 2 }, FIELD_ID3, FIELD_ID2);
    assertEquals(asList(null, 2, 1), r.intoList());
}
```

Which fails as follows:

```
java.lang.AssertionError: expected:<[null, 1, null]> but was:<[null, null, null]>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at org.jooq.test.AbstractTest.assertEquals(AbstractTest.java:194)
	at org.jooq.test.RecordUnmappingTest.testFromArrayExplicitFields(RecordUnmappingTest.java:102)
```

**Comment by lukaseder** [comment]
The issue is also present in 3.16.5 and 3.15.9

**Comment by lukaseder** [comment]
Confirmed, the test code worked in 3.15.5 and no longer works in 3.15.6 due to https://github.com/jOOQ/jOOQ/issues/12704

**Comment by lukaseder** [comment]
Even before the regression introduced by #12704, the implementation loading lists into a record was inconsistently different from the one loading arrays. Before it was removed, the method `AbstractRecord::prepareArrayForUnmap` patched only arrays, not lists:
https://github.com/jOOQ/jOOQ/blob/version-3.15.5/jOOQ/src/main/java/org/jooq/impl/AbstractRecord.java#L918

That method needs to be restored, and applied to both `Object[]` and `Iterable` types to introduce the necessary gaps in the array/iterable prior to loading it into the intermediary `Record` from which the `this` record can then be loaded.

**Comment by alf** [comment]
Does this mean that using a List instead of an array is a valid workaround?

**Comment by lukaseder** [comment]
I don't think so

**Comment by lukaseder** [comment]
As a workaround, you can implement an `UnmapperProvider`

**Comment by alf** [comment]
Thanks, it seems this will be fixed pretty soon so we’ll probably just wait :)

**Comment by lukaseder** [comment]
Yeah, probably better. You can always just loop over the array and set each value individually, too.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0, 3.16.6 (https://github.com/jOOQ/jOOQ/issues/13279), and 3.15.10 (https://github.com/jOOQ/jOOQ/issues/13280)
